### PR TITLE
Talwai/agent dev mode

### PR DESCRIPTION
**Pull request by talwai**
This adds a "developer mode" to the agent, that can be enabled by the `--profile` command line flag or by the `developer_mode` setting in `agentConfig`. 

Developer mode can be enabled at the `check` level or the `collector` level. When enabled at the check
level, e.g. `./agent.py check nginx --profile`, the `check.run()` function is profiled with `cProfile` and the `pstats` output is dumped to the log. 

When enabled at the `collector` level, the following behavior is enabled:
- Every run of the collector loop is profiled with `cProfile`, in the same way as above
- The `AgentMetrics` check (`checks.d/agent_metrics.py`) is run at the end of each collector loop. This check can be configured to run a set of additional `psutil.Process` methods on the current process. These additional metrics are flushed to the dispatcher as well as dumped to the log. See `conf.d/agent_metrics.yaml.default` for the default configuration, which can be extended as needed. Currently a setting under `process_metrics` is ignored if there is no corresponding `psutil.Process` method.

### CODE REVIEW

**Comment by degemer** [comment]
Hey @talwai, I tested it a little and have some remarks (not concerning the code) about it, you probably already talked about it with @LeoCavaille, but I'm going to add my 2 cents.

I was expecting a lot more "by check" metrics, ie at the end of every run, display some stats for each check (and not only globally), for instance run time, memory consumption after/before, io read/write (basically the metrics you're already displaying for each run, and maybe tag them by check if you want to upload them to DD).
Also, about these metrics, could you display them in a nicer way ? Sure, once it's uploaded to DD, it should be really easy to analyze them, but a display in the logs would be a real plus.
Actually I was thinking something like the `info` display. :smile: (`collector` being the "global" metrics)
```bash
consul:
  run_time: 1.01s
  memory_before: 60785934
  memory_after: 60785935
  io read: 2
  ...
pgbouncer:
  run_time: 6.61s
  memory_before: 60785834
  memory_after: 60785840
  io read: 255
  ...
collector:
  run_time: 15.04s
  memory_before: 60785834
  memory_after: 60786001
  io read: 256
  ...
```

If you do something like this, I think that the profiler should be another option, not activated by default with the "developer mode", because we don't always need this level of details when debugging a check. And maybe also a profiler by check ? (not sure it's needed, just a suggestion)

Another thing: I don't think you should put `agent_metrics.py` in `checks.d`, because when you do this it becomes visible in 
> initialized checks.d checks: ['agent_metrics', 'network', 'pgbouncer', 'ntp', 'consul', 'nginx']

and a standard user shouldn't see this (event if it doesn't do anything).

The new agent_metrics are really nice! :+1: I think you could just prefix them by `datadog.agent.collector` instead of `datadog.agent`. (and as you wrote in your `TODOs`, some are probably rates and not gauges)

I just realized that we already have some agent_metrics here: https://github.com/DataDog/dd-agent/blob/master/checks/agent_metrics.py, you might want to put yours there too. (or maybe you split them on purpose ?)

**Comment by talwai** [comment]
@degemer Thanks for the feedback!
> I was expecting a lot more "by check" metrics, ie at the end of every run, display some stats for each check (and not only globally), for instance run time, memory consumption after/before, io read/write 

I agree that we should be getting more data about the performance of individual checks, and not just the collector run as a whole. I can extend `wrap_profiling` to include some before and after `psutil` calls. Though I wonder if the extra system calls would hinder the performance of the check itself i.e. `check.run()` becomes `psutil.get_memory_info() ; check.run() ; psutil.get_memory_info()` for every check. 
> Also, about these metrics, could you display them in a nicer way ? 

Yep, the output is ugly. Will improve the formatting for the log dumps

> I think that the profiler should be another option, not activated by default with the "developer mode", because we don't always need this level of details when debugging a check. And maybe also a profiler by check ? (not sure it's needed, just a suggestion)

I get where you're coming from. In fact, I myself silenced the profiler dump when debugging the `AgentMetrics` check because it was mostly unhelpful. I wonder if that is the case in general though. Easy enough to make them two separate options, but would be interested in hearing other opinions.

> Another thing: I don't think you should put agent_metrics.py in checks.d, because when you do this it becomes visible in
initialized checks.d checks: ['agent_metrics', 'network', 'pgbouncer', 'ntp', 'consul', 'nginx']

The reason it's in there is: 
1. Because it appears that that's where all the new-style checks go 
2. So that it can be tested via `AgentCheckTest`. 
I realize that it is possible to test without the scaffolding, but my thought was that all tests for new-style checks should follow the same pattern. If the only issue with locating it in `checks.d/` is the log output, then I would prefer to just special-case it so it doesn't show up. Admittedly there may be other issues with having it in `checks.d/` that I'm not aware of.

> I just realized that we already have some agent_metrics here: https://github.com/DataDog/dd-agent/blob/master/checks/agent_metrics.py, you might want to put yours there too. (or maybe you split them on purpose ?)

Basically `checks.d/agent_metrics.py` is all of `checks/agent_metrics.py` and more. I opted against simply extending the class in place because it uses the old `Check` API, and if we plan to deprecate that API we should probably stop using it internally.

File: utils/profile.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#3p
+import psutil
```
Comment (LeoCavaille) [inline utils/profile.py]: You probably don't need psutil here!?

File: utils/profile.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import os
+import logging
+from config import _is_affirmative
```
Comment (LeoCavaille) [inline utils/profile.py]: This also can go away.

File: checks/__init__.py [hunk 3]
```diff
@@ -531,8 +541,43 @@ def get_warnings(self):
         self.warnings = []
         return warnings
 
+    @staticmethod
+    def _collect_stats():
+        process = psutil.Process(os.getpid())
+        mem_info = process.get_memory_info()._asdict()
+        io_info = process.get_io_counters()._asdict()
```
Comment (LeoCavaille) [inline checks/__init__.py]: This one is not available on OS X for instance :/

File: checks/__init__.py [hunk 4]
```diff
@@ -560,6 +607,14 @@ def run(self):
...
                 )
             instance_statuses.append(instance_status)
+
+        if self.in_developer_mode:
+            try:
+                after = AgentCheck._collect_stats()
+                self.set_stats(before, after)
+            except Exception: # It's fine if we can't collect stats for the run
+                pass
+
```
Comment (LeoCavaille) [inline checks/__init__.py]: Here I am not sure if for instance the CPU percentage sampling before and after would be useful (same for IO) because we would need to see it _during_ the call of `.check()`
For memory, we could see the leaks I guess.

File: conf.d/agent_metrics.yaml.default [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+        get_io_counters: yes
+        get_num_threads: yes
+        get_connections: no
+    general:
+        foo: bar
+
+instances:
+  # No configuration is needed for this check.
+  # A single instance needs to be defined with any value.
+  - foo: bar
```
Comment (LeoCavaille) [inline conf.d/agent_metrics.yaml.default]: `[{}]`

File: checks.d/agent_metrics.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+import psutil
+import os
+import threading
+from checks import AgentCheck
+from config import _is_affirmative
```
Comment (LeoCavaille) [inline checks.d/agent_metrics.py]: This should be ordered `#stdlib #3p #project`

File: checks.d/agent_metrics.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                method_key = method[4:] if method.startswith('get_') else method
+                try:
+                    raw_stats = getattr(current_process, method)()
+                    try:
+                        stats[method_key] = raw_stats._asdict()
+                    except AttributeError:
+                        if isinstance(raw_stats, int):
+                            stats[method_key] = raw_stats
+                        else:
+                            self.log.warn("Could not serialize output of {} to dict".format(method))
```
Comment (LeoCavaille) [inline checks.d/agent_metrics.py]: The `{}` syntax is not python 2.6 compatible... It should be named placeholders `{0}`.

**Comment by LeoCavaille** [comment]
Nice! Good to see progress, this will prove super helpful to help the community design more performant checks. Made a few remarks just giving a first look.
Two ideas of stuff that it would be super cool to have:
* rather than just logging a dozen lines of pstats in logs at every run (and sometimes it might not be exactly what you want actually), could it be possible to enable profiling at a higher level for a while and accumulate the whole collector pstats?
* and optionally to be able to dump that to a file instead of logs so that you can use pstats for forensics later on offline.
What do you think?

File: checks/__init__.py [hunk 8]
```diff
@@ -22,6 +22,7 @@
 
 # 3rd party
 import yaml
+import psutil
```
Comment (remh) [inline checks/__init__.py]: psutil is actually not available on source install without a compiler :/ 
Can you try and catch an import error and disable the developer mode accordingly ?

**Comment by talwai** [comment]
@LeoCavaille thanks for the review! I think both your ideas are good, if I understand correctly. I will look into moving the profiler higher in the call stack, and dumping cumulative statistics after longer intervals, rather than on every collector run. This should give us more useful higher-level insight than the current setup does. I will also add an option for dumping `pstats` to a file instead of / in addition to logging.

**Comment by talwai** [comment]
The updated default functionality is as follows:

1. `agent.py` developer mode: Enables more granular checks in `checks.d/agent_metrics.py`. These checks are flushed to Datadog and also dumped to `log.info`. A `pstats` profile is dumped to the file `./collector-stats.dmp` after `COLLECTOR_PROFILE_INTERVAL` runs ( both the file name and the interval are candidates for moving to `agentConfig`). Additionally every check calls the static method `AgentCheck._collect_stats` before and after its run. This collects some `psutil` stats, the defaults of which are specified here: https://github.com/DataDog/dd-agent/blob/talwai/agent_dev_mode/checks/__init__.py#L37 (maybe also a candidate for living in `agentConfig`). These stats are prettified and dumped to `log.info`

2. `check` developer mode: This collects pstats output of the check run and dumps to `log.debug`. (Maybe it can dump to a file as well? Seems overkill). Additionally it calls `AgentCheck._collect_stats()` before and after its run and outputs data in the `agent.py info` format. e.g.: 
```
$ python agent.py check ntp --profile
2015-04-30 11:20:44,471 | INFO | dd.collector | config(config.py:922) | initialized checks.d checks: ['ntp', 'network']
2015-04-30 11:20:44,471 | INFO | dd.collector | config(config.py:923) | initialization failed checks.d checks: []
2015-04-30 11:20:44,471 | INFO | dd.collector | checks.collector(collector.py:435) | Running check ntp
Metrics:  [('ntp.offset', 1430407244, 0.02178668975830078, {'hostname': 'Aadityas-MacBook-Pro.local', 'type': 'gauge'})]
Events:  []
Service Checks:  [{'status': 0, 'tags': None, 'timestamp': 1430407244.5470452, 'check': 'ntp.in_sync', 'host_name': 'Aadityas-MacBook-Pro.local', 'message': None, 'id': 1}]
    ntp
    ---
      - instance #0 [OK] Last run duration: 0.0944309234619
      - Collected 1 metric, 0 events & 1 service check
      - Stats: 
            Memory Before (RSS): 18092032
            Memory After (RSS): 18108416
            Memory Before (VMS): 2534723584
            Memory After (VMS): 2534723584
```

File: checks/__init__.py [hunk 9]
```diff
@@ -13,18 +14,30 @@
...
 from config import get_confd_path
 from checks import check_status
+from checks.utils import pretty_statistics
 
 # 3rd party
 import yaml
 
+try:
+    import psutil
+    PSUTIL_PRESENT = True
```
Comment (remh) [inline checks/__init__.py]: You don't need this flag.
To test if psutil is here you can just do 
```python
if psutil is not None
```

Comment (remh) [inline checks/__init__.py]: e.g. https://github.com/DataDog/dd-agent/blob/1bef95f968c1d884e6aeddc93a692d8a387535c8/checks/system/win32.py#L5-L8

File: agent.py [hunk 10]
```diff
@@ -50,6 +52,11 @@
 START_COMMANDS = ['start', 'restart', 'foreground']
 DD_AGENT_COMMANDS = ['check', 'flare', 'jmx']
 
+PSTATS_LIMIT = 20
```
Comment (remh) [inline agent.py]: Should those constants be in a different module ?

File: agent.py [hunk 11]
```diff
@@ -134,19 +143,22 @@ def run(self, config=None):
...
-                    profiler.disable()
-                    import pstats
-                    from cStringIO import StringIO
-                    s = StringIO()
-                    ps = pstats.Stats(profiler, stream=s).sort_stats("cumulative")
-                    ps.print_stats()
-                    log.debug(s.getvalue())
-                except Exception:
-                    log.warn("Cannot disable profiler")
+            if profiled:
```
Comment (remh) [inline agent.py]: Can we move this code to the new module ?

File: checks/__init__.py [hunk 12]
```diff
@@ -265,6 +278,12 @@ def get_metrics(self, expire=True):
                 pass
         return metrics
 
+    def get_stats(self):
```
Comment (remh) [inline checks/__init__.py]: Do we need it here? This class is a legacy class, not sure it's worth any further development.

File: checks/__init__.py [hunk 13]
```diff
@@ -531,8 +553,61 @@ def get_warnings(self):
...
+            except psutil.AccessDenied:
+                if log is not None:
+                    log.warn("Cannot call psutil method {} : Access Denied".format(method))
+
+        return stats
+
+    def set_stats(self, before, after):
+        self.stats = {'before': before, 'after': after}
+
+    def get_stats(self):
```
Comment (remh) [inline checks/__init__.py]: Watchout:
```python
get_stats
```
name 
is likely to be used inside checks.d.
It's already used by https://github.com/DataDog/dd-agent/blob/master/checks.d/rabbitmq.py#L162
And is likely used by some custom checks.d. 

You should use something different such as:
```python
def _get_internal_profiling_stats(...):
 ```

File: checks/__init__.py [hunk 14]
```diff
@@ -531,8 +553,61 @@ def get_warnings(self):
...
+        stats = self.stats
+        self.stats = None
+        return stats
+
     def run(self):
         """ Run all instances. """
+
+        # Store run statistics if needed
+        before, after = None, None
+        if self.in_developer_mode and self.name != 'agent_metrics':
```
Comment (remh) [inline checks/__init__.py]:
```
agent_metrics
```
should be a constant

File: checks/__init__.py [hunk 15]
```diff
@@ -560,6 +649,15 @@ def run(self):
...
                 )
             instance_statuses.append(instance_status)
+
+        if self.in_developer_mode and self.name != 'agent_metrics':
+            try:
+                after = AgentCheck._collect_stats()
+                self.set_stats(before, after)
+                log.info("\n \t %s %s" % (self.name, pretty_statistics(self.stats)))
+            except Exception: # It's fine if we can't collect stats for the run
+                pass
```
Comment (remh) [inline checks/__init__.py]: we should probably log something in debug ?

File: checks/collector.py [hunk 16]
```diff
@@ -390,6 +430,50 @@ def run(self, checksd=None, start_event=True):
...
+            current_service_checks = check.get_service_checks()
+
+            check_stats = check.get_stats()
+
+            # Save the status of the check.
+            metric_count = len(current_check_metrics)
+            event_count = len(current_check_events)
+            service_check_count = len(current_service_checks)
+
+            import pprint
```
Comment (remh) [inline checks/collector.py]: already imported

File: checks/utils.py [hunk 17]
```diff
@@ -116,3 +116,24 @@ def tail(self, line_by_line=True, move_end=True):
...
+
+    before = stats.get('before')
+    after = stats.get('after')
+
+    mem_before = before.get('memory_info')
+    mem_after = after.get('memory_info')
+
+    if mem_before and mem_after:
+        return """
+            Memory Before (RSS): {0}
```
Comment (remh) [inline checks/utils.py]: This should probably display the difference after - before

File: config.py [hunk 18]
```diff
@@ -56,12 +56,14 @@
     'nagios_perf_cfg'
     ]
 
-DEFAULT_CHECKS = ("network", "ntp")
```
Comment (remh) [inline config.py]: :+1:

**Comment by remh** [comment]
Great work!
We should probably add the option in the datadog.conf by the way.

Also does it work properly on windows?

**Comment by talwai** [comment]
Great question on Windows - I'm inclined to say that it works wherever `psutil` and `time.time()` work but will try it on a VM to be sure

**Comment by talwai** [comment]
@remh Suggestions implemented. B2U

File: checks/collector.py [hunk 19]
```diff
@@ -144,14 +149,26 @@ def run(self, checksd=None, start_event=True):
...
         service_checks = payload['service_checks']
-        
+        if checksd:
+            self.initialized_checks_d = checksd['initialized_checks'] # is of type {check_name: check}
+            self.init_failed_checks_d = checksd['init_failed_checks'] # is of type {check_name: {error, traceback}}
+
+            # Find the AgentMetrics check and pop it out
+            # This check must run at the end of the loop to collect info on agent performance
+            if not self._agent_metrics:
+                for check in self.initialized_checks_d[:]:
```
Comment (remh) [inline checks/collector.py]: you don't need the:
```python
[:]
```

File: agent.py [hunk 20]
```diff
@@ -117,36 +123,30 @@ def run(self, config=None):
...
+                if collector_profiled_runs >= COLLECTOR_PROFILE_INTERVAL:
+                    try:
+                        profiler.disable_profiling()
+                        profiled = False
+                        collector_profiled_runs = 0
+                    except Exception:
```
Comment (remh) [inline agent.py]: Any idea what kind of exception can be raised here ? 
If you just want to play it safe to make sure that the profiler doesn't crash the agent (which is a good thing) can you log the raised exception please ?

File: agent.py [hunk 21]
```diff
@@ -50,6 +52,9 @@
 START_COMMANDS = ['start', 'restart', 'foreground']
 DD_AGENT_COMMANDS = ['check', 'flare', 'jmx']
 
+# How many runs to profile before dumping pstats
+COLLECTOR_PROFILE_INTERVAL = 1
```
Comment (remh) [inline agent.py]: Should that be configurable ?

File: checks/__init__.py [hunk 22]
```diff
@@ -286,6 +300,9 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.name = name
         self.init_config = init_config or {}
         self.agentConfig = agentConfig
+        self.in_developer_mode = agentConfig.get('developer_mode') and psutil is not None
+        self.stats = None
```
Comment (remh) [inline checks/__init__.py]: Same here, you should use a name that have less chances to conflict with existing checks

File: checks.d/agent_metrics.py [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+    def set_metric_context(self, payload, context):
+        self._collector_payload = payload
+        self._metric_context = context
+
+    def get_metric_context(self):
+        return self._collector_payload, self._metric_context
+
+    def check(self, instance):
+        in_developer_mode = self.agentConfig.get('developer_mode', False)
+        if in_developer_mode and psutil is not None:
```
Comment (remh) [inline checks.d/agent_metrics.py]: This is redundant with your 
```python
self.in_developer_mode
```
https://github.com/DataDog/dd-agent/blob/talwai/agent_dev_mode/checks/__init__.py#L303

**Comment by talwai** [comment]
@remh This works on windows

File: checks/__init__.py [hunk 24]
```diff
@@ -564,6 +648,15 @@ def run(self):
                     tb=traceback.format_exc()
                 )
             instance_statuses.append(instance_status)
+
+        if self.in_developer_mode and self.name != 'agent_metrics':
```
Comment (remh) [inline checks/__init__.py]: nitpick but you should use your 
```python
AGENT_METRICS_CHECK_NAME
```
constant

File: checks.d/agent_metrics.py [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+         }
+
+         This creates a metric like `datadog.agent.collector.{key_1}.{key_2}` where key_1 is a top-level
+         key in `stats`, and key_2 is a nested key.
+         E.g. datadog.agent.collector.memory_info.rss
+        """
+
+        base_metric = 'datadog.agent.collector.{0}.{1}'
+        #TODO: May have to call self.normalize(metric_name) to get a compliant name
+        for k, v in stats.items():
```
Comment (remh) [inline checks.d/agent_metrics.py]: You should use an iterator
```
.iteritems()
```

File: checks.d/agent_metrics.py [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+
+        :returns a dictionary of statistic_name: value
+        """
+        process_metrics = instance.get('process_metrics', self.init_config.get('process_metrics', None))
+        if not process_metrics:
+            self.log.error('No metrics configured for AgentMetrics check!')
+            return {}
+
+        methods, metric_types = zip(
+            *[(p['name'], p.get('type', 'gauge'))
```
Comment (remh) [inline checks.d/agent_metrics.py]: Nitpick but you should use constants for the metric types.

They are actually already existing:
https://github.com/DataDog/dd-agent/blob/master/checks/metric_types.py

File: checks/__init__.py [hunk 27]
```diff
@@ -535,8 +552,61 @@ def get_warnings(self):
         self.warnings = []
         return warnings
 
+    @staticmethod
+    def _get_statistic_name_from_method(method_name):
+        return method_name[4:] if method_name.startswith('get_') else method_name
+
+    @staticmethod
+    def _collect_stats(methods=None, log=None):
```
Comment (remh) [inline checks/__init__.py]: You should find a better name:
https://github.com/DataDog/dd-agent/blob/master/checks.d/postgres.py#L351
https://github.com/DataDog/dd-agent/blob/master/checks.d/pgbouncer.py#L69

What about 
```python
_collect_internal_stats
```
?

File: checks/__init__.py [hunk 28]
```diff
@@ -535,8 +552,61 @@ def get_warnings(self):
...
+            stat_name = AgentCheck._get_statistic_name_from_method(method)
+            try:
+                raw_stats = getattr(current_process, method)()
+                try:
+                    stats[stat_name] = raw_stats._asdict()
+                except AttributeError:
+                    if isinstance(raw_stats, numbers.Number):
+                        stats[stat_name] = raw_stats
+                    else:
+                        if log is not None:
```
Comment (remh) [inline checks/__init__.py]: When would log be None ?

We should always have a logger set in this code path.

**Comment by remh** [comment]
Looks great beside the last comments!

It should be ready to merge once those are fixed

File: config.py [hunk 29]
```diff
@@ -57,12 +57,14 @@
...
 ]
 
-DEFAULT_CHECKS = ("network", "ntp")
 LEGACY_DATADOG_URLS = [
     "app.datadoghq.com",
     "app.datad0g.com",
 ]
 
+#Checks whose log output to suppress, unless explicitly asked for
+HIDDEN_CHECKS = ['agent_metrics']
```
Comment (remh) [inline config.py]: You should use the constant you defined here.

File: tests/checks/mock/test_agent_metrics.py [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+
+MOCK_NAMES_TO_METRIC_TYPES = {
+    'memory_info': 'gauge',
+    'cpu_times': 'gauge'
+}
+
+
+class AgentMetricsTestCase(AgentCheckTest):
+
+    CHECK_NAME = 'agent_metrics'
```
Comment (remh) [inline tests/checks/mock/test_agent_metrics.py]: Same here

File: checks/__init__.py [hunk 31]
```diff
@@ -13,18 +14,30 @@
...
+    import psutil
+except ImportError:
+    psutil = None
+
 log = logging.getLogger(__name__)
 
+# Default methods run when collecting info about the agent in developer mode
+DEFAULT_PSUTIL_METHODS = ['get_memory_info', 'get_io_counters']
+
+AGENT_METRICS_CHECK_NAME = 'agent_metrics'
```
Comment (remh) [inline checks/__init__.py]: You shouldn't define it twice
